File: typescript/robot-simulator/robot-simulator.ts
export default class Robot {
  public bearing: string;
  public coordinates: [number, number];

  constructor(x: number = 0, y: number = 0, direction: string = "north") {
    this.bearing = direction;
    this.coordinates = [x, y];
  }

  public orient(bearing: string) {
    this.bearing = bearing;
  }

  public at(x: number, y: number) {
    this.coordinates = [x, y];
  }

  public turnRight() {
    switch (this.bearing) {
      case "north":
        this.bearing = "east";
        break;
      case "east":
        this.bearing = "south";
        break;
      case "south":
        this.bearing = "west";
        break;
      case "west":
        this.bearing = "north";
        break;
    }
  }

  public turnLeft() {
    switch (this.bearing) {
      case "north":
        this.bearing = "west";
        break;
      case "east":
        this.bearing = "north";
        break;
      case "south":
        this.bearing = "east";
        break;
      case "west":
        this.bearing = "south";
        break;
    }
  }

  public advance() {
    const [x, y] = this.coordinates;

    switch (this.bearing) {
      case "north":
        this.coordinates = [x, y + 1];
        break;
      case "east":
        this.coordinates = [x + 1, y];
        break;
      case "south":
        this.coordinates = [x, y - 1];
        break;
      case "west":
        this.coordinates = [x - 1, y];
        break;
    }
  }

  public instructions(input: string) {
    return [...input].map(instruction => this.getInstruction(instruction));
  }

  private getInstruction(instruction: string) {
    switch (instruction) {
      case "L":
        return "turnLeft";
      case "R":
        return "turnRight";
      case "A":
        return "advance";
      default:
        throw new Error("Invalid instruction");
    }
  }

  public evaluate(instructions: string) {
    [...instructions].map(instruction => this.evaluateInstruction(instruction));
  }

  private evaluateInstruction(instruction: string) {
    switch (instruction) {
      case "L":
        this.turnLeft();
        break;
      case "R":
        this.turnRight();
        break;
      case "A":
        this.advance();
        break;
      default:
        throw new Error("Invalid instruction");
    }
  }
}

File: ruby/robot-simulator/robot_simulator.rb
class Robot
  attr_accessor :coordinates, :bearing

  def initialize
    @coordinates = [0, 0]
    @bearing = :north
  end

  def orient(direction)
    raise ArgumentError, 'Invalid direction' unless BEARINGS.include?(direction)

    @bearing = direction
  end

  def at(x, y) = @coordinates = [x, y]

  def turn_right
    case @bearing
    when :north then @bearing = :east
    when :east  then @bearing = :south
    when :south then @bearing = :west
    when :west  then @bearing = :north
    end
  end

  def turn_left
    case @bearing
    when :north then @bearing = :west
    when :east  then @bearing = :north
    when :south then @bearing = :east
    when :west  then @bearing = :south
    end
  end

  def advance
    case @bearing
    when :north then @coordinates[1] += 1
    when :east  then @coordinates[0] += 1
    when :south then @coordinates[1] -= 1
    when :west  then @coordinates[0] -= 1
    end
  end

  BEARINGS = %i[east west north south].freeze
  private_constant :BEARINGS
end

class Simulator
  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, word)
    instructions(word).each_with_object(robot) do |instruction, evaluating_robot|
      evaluating_robot.send(instruction)
    end
  end

  def instructions(word) = word.each_char.map { |letter| LETTER_TO_INSTRUCTION[letter] }

  LETTER_TO_INSTRUCTION = { 'A' => :advance, 'L' => :turn_left, 'R' => :turn_right }.freeze
  private_constant :LETTER_TO_INSTRUCTION
end